storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+#define SD_FREQ_25M     25000000 /* 25 MHz clock usable when card is initialized */
+#define SD_FREQ_50M     50000000 /* 50 MHz clock usable when card is initialized and supports high speed */
+
+#define ARRAY_LENGTH(array) (sizeof(array) / sizeof(array[0]))
+
+typedef struct {
+	/* Relative Card Address of the card. Needed for selecting the card so it can use the data bus. */
+	uint32_t rca;
+	/* Size of SD card in blocks */
+	uint32_t sizeBlocks;
+	/* Size (in blocks) of an erase sector */
+	uint32_t eraseSizeBlocks;
+	/* Number of command timeouts since last card initialization */
+	uint32_t commandTimeouts;
+	/* Whether the card supports SDHC protocol (requires slightly different handling) */
+	bool highCapacity;
+} sdcard_cardMetadata_t;
+
+typedef struct {
+	/* Base address (in virtual memory) of SD Host Controller register set */
+	volatile uint32_t *base;
+	handle_t cmdLock;
+	sdcard_cardMetadata_t card;
+	/* Reference clock frequency in Hz */
+	uint32_t refclkFrequency;
+
+	/* Address of DMA buffer in virtual memory (for access by the CPU) */
+	void *dmaBuffer;
+	/* Address of DMA buffer in physical memory (for access by the SD Host Controller) */
+	addr_t dmaBufferPhys;
+
+	bool sdioInitialized;
+	bool isCDPinSupported;
+	bool isWPPinSupported;
+
+	handle_t isrHandle;
+	handle_t eventLock;
+	handle_t eventCond;
+} sdcard_hostData_t;
+
+static sdcard_hostData_t sdio_hosts[PLATFORM_SDIO_N_HOSTS] = { 0 };
+static unsigned int initializedHosts = 0;
+
+static struct {
+	handle_t lock;
+	handle_t cond;
+	bool initialized;
+} presenceEvents = { .initialized = false };
+
+static inline sdcard_hostData_t *sdcard_getHostForSlot(unsigned int slot)
+{
+	return (slot >= PLATFORM_SDIO_N_HOSTS) ? NULL : &sdio_hosts[slot];
+}
+
+#define GET_HOST_OR_RETURN(...) \
+	sdcard_hostData_t *host = sdcard_getHostForSlot(slot); \
+	if (host == NULL) { \
+		return __VA_ARGS__; \
+	}
+
+static int sdcard_configClockAndPower(sdcard_hostData_t *host, uint32_t freq);
+static int sdcard_wideAndFast(sdcard_hostData_t *host);
+
+
+/* Resets parts or all of the host according to the given reset type.
+ * Argument must be one of CLOCK_CONTROL_RESET_*
+ */
+static int sdhost_reset(sdcard_hostData_t *host, uint32_t resetType)
+{
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
+	for (int i = 0; i < SDHOST_RETRIES; ++i) {
+		usleep(10);
+		uint32_t val = *(host->base + SDHOST_REG_CLOCK_CONTROL) & resetType;
+		if (val == 0) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+
+static inline bool sdcard_isWriteProtected(sdcard_hostData_t *host)
+{
+	if (host->isWPPinSupported) {
+		uint32_t val = *(host->base + SDHOST_REG_PRES_STATE) & PRES_STATE_WRITE_PROT_PIN;
+		if (val == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+static void *sdhost_mmapDMA(size_t size)
+{
+	size_t sz = (size + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
+	void *p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, NULL, 0);
```

Comment:
Alas `mmap` API changed in the meantime
```suggestion
	void *p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, -1, 0);
```